### PR TITLE
 Add typed transform inputs

### PR DESCRIPTION
**Pull request by narape**
# TL;DR
This feature allows the users to know the specific inputs from a task at development time. This is really important when you are working with a remote task that is developed by other users.

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [ ] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
This feature allows the users to know the specific inputs from a task at development time. This is really important when you are working with a remote task that is developed by other users. The final goal is to use the input class in the WorkflowBuilder instead of passing the input into the SdkTransform using the withInput(...) method that forces you to know the attribute names of the input class. Example: [Fibonacci Workflow](https://github.com/flyteorg/flytekit-java/blob/5a8aea9be56da9c2098e048cc08b1398468ddec6/flytekit-examples/src/main/java/org/flyte/examples/FibonacciWorkflow.java#L46-L51)

## Tracking Issue
* https://github.com/flyteorg/flyte/issues/3260

## Follow-up issue
_NA_

### CODE REVIEW

File: flytekit-examples-scala/src/main/scala/org/flyte/examples/flytekitscala/AddQuestionTask.scala [hunk 1]
```diff
@@ -46,18 +46,3 @@ class AddQuestionTask
...
-object AddQuestionTask {
-
-  /** Binds input data to this task
-    *
-    * @param greeting
-    *   the input greeting message
-    * @return
-    *   a transformed instance of this class with input data
-    */
-  def apply(
```
Comment (narape) [inline flytekit-examples-scala/src/main/scala/org/flyte/examples/flytekitscala/AddQuestionTask.scala:59]: No more `apply` or `of` methods in Scala or Java

File: flytekit-examples/src/main/java/org/flyte/examples/AddQuestionTask.java [hunk 2]
```diff
@@ -34,23 +33,17 @@ public AddQuestionTask() {
...
-
   /**
    * Generate an immutable value class that represents {@link AddQuestionTask}'s input, which is a
    * String.
    */
   @AutoValue
   public abstract static class Input {
     public abstract SdkBindingData<String> greeting();
+
+    public static Input create(SdkBindingData<String> greeting) {
```
Comment (narape) [inline flytekit-examples/src/main/java/org/flyte/examples/AddQuestionTask.java:44]: Before we were only using the AutoValue to define the type of the task, but now we actually create the auto values to pass it in the apply

File: flytekit-java/src/main/java/org/flyte/flytekit/SdkCondition.java [hunk 3]
```diff
@@ -18,42 +18,63 @@
 
 import java.util.ArrayList;
 import java.util.List;
-import java.util.Map;
 import javax.annotation.Nullable;
 
-public class SdkCondition<OutputT> extends SdkTransform<OutputT> {
+public class SdkCondition<OutputT> extends SdkTransform<Void, OutputT> {
```
Comment (narape) [inline flytekit-java/src/main/java/org/flyte/flytekit/SdkCondition.java:23]: **Note for reviewers**

We made a SdkCondition to have no inputs on itself but the transformations going on when, and otherwise cases could have a combination of both:
* SdkTransforms that takes *no* inputs
* SdkTransforms that takes inputs but the inputs are supplied

All of them should have the same ouput type

File: flytekit-java/src/main/java/org/flyte/flytekit/SdkPartialTransform.java [hunk 4]
```diff
@@ -16,72 +16,42 @@
...
 import java.util.List;
-import java.util.Map;
 import java.util.function.Function;
 import javax.annotation.Nullable;
 
 /** {@link SdkTransform} with partially specified inputs. */
-class SdkPartialTransform<T> extends SdkTransform<T> {
-  private final SdkTransform<T> transform;
-  private final Map<String, SdkBindingData<?>> fixedInputs;
+class SdkPartialTransform<InputT, OutputT> extends SdkTransform<InputT, OutputT> {
```
Comment (narape) [inline flytekit-java/src/main/java/org/flyte/flytekit/SdkPartialTransform.java:30]:
```suggestion
class SdkTransformMetadataDecorator<InputT, OutputT> extends SdkTransform<InputT, OutputT> {
```
Reading the java doc made me realize that the name is no longer relevant, given that we no longer hold inputs in the wrapper class. We need to rename the class and change the now obsolete doc

File: flytekit-java/src/main/java/org/flyte/flytekit/SdkRemoteLaunchPlan.java [hunk 5]
```diff
@@ -107,7 +115,7 @@ public SdkNode<OutputT> apply(
         WorkflowNode.builder()
             .reference(WorkflowNode.Reference.ofLaunchPlanRef(workflowId))
             .build(),
-        inputs,
+        inputsBindings,
```
Comment (narape) [inline flytekit-java/src/main/java/org/flyte/flytekit/SdkRemoteLaunchPlan.java]: Revert this change

File: flytekit-java/src/main/java/org/flyte/flytekit/SdkAppliedTransform.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.flyte.flytekit;
+
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+class SdkAppliedTransform<OriginalInputT, OutputT> extends SdkTransform<Void, OutputT> {
```
Comment (narape) [inline flytekit-java/src/main/java/org/flyte/flytekit/SdkAppliedTransform.java:23]: Add Javadoc here

File: flytekit-java/src/main/java/org/flyte/flytekit/SdkTransform.java [hunk 7]
```diff
@@ -16,72 +16,55 @@
...
-public abstract class SdkTransform<T> {
+public abstract class SdkTransform<InputT, OutputT> {
 
-  public abstract SdkType<T> getOutputType();
+  public abstract SdkType<InputT> getInputType();
 
-  public abstract SdkNode<T> apply(
+  public abstract SdkType<OutputT> getOutputType();
+
+  public final SdkNode<OutputT> apply(
```
Comment (narape) [inline flytekit-java/src/main/java/org/flyte/flytekit/SdkTransform.java:34]: We should add unit tests that exercise the `checkNullOnlyVoid` method

File: flytekit-java/src/main/java/org/flyte/flytekit/SdkTransform.java [hunk 8]
```diff
@@ -91,4 +74,16 @@ public SdkTransform<T> withTimeoutOverride(Duration timeout) {
   public String getName() {
     return getClass().getName();
   }
+
+  void checkNullOnlyVoid(@Nullable InputT inputs) {
```
Comment (narape) [inline flytekit-java/src/main/java/org/flyte/flytekit/SdkTransform.java:78]: This method makes sure that the input "nullness" matches the input type:
* `null` in the only valid value for `Void` and `Unit` `SdkType`s
* Any other `SdkType` cannot have `null`

If you have a better name please suggest it

File: flytekit-java/src/test/java/org/flyte/flytekit/SdkLaunchPlanRegistrarTest.java [hunk 9]
```diff
@@ -55,11 +54,7 @@ class SdkLaunchPlanRegistrarTest {
   private static final Map<String, String> ENV;
 
   static {
-    HashMap<String, String> env = new HashMap<>();
-    env.put(PROJECT_ENV_VAR, "project");
-    env.put(DOMAIN_ENV_VAR, "domain");
-    env.put(VERSION_ENV_VAR, "version");
-    ENV = Collections.unmodifiableMap(env);
+    ENV = Map.of(PROJECT_ENV_VAR, "project", DOMAIN_ENV_VAR, "domain", VERSION_ENV_VAR, "version");
```
Comment (narape) [inline flytekit-java/src/test/java/org/flyte/flytekit/SdkLaunchPlanRegistrarTest.java:57]: No need of static initializer, assign all the values in line 54

File: flytekit-local-engine/src/test/java/org/flyte/localengine/examples/TestUnaryIntegerOutput.java [hunk 10]
```diff
@@ -33,36 +26,4 @@ public abstract class TestUnaryIntegerOutput {
   public static TestUnaryIntegerOutput create(SdkBindingData<Long> o) {
     return new AutoValue_TestUnaryIntegerOutput(o);
   }
-
-  public static class SdkType extends org.flyte.flytekit.SdkType<TestUnaryIntegerOutput> {
```
Comment (narape) [inline flytekit-local-engine/src/test/java/org/flyte/localengine/examples/TestUnaryIntegerOutput.java:37]: Note to reviewers, custom `SdkType` implementation is not necessary here as we can use the Jackson one
